[DOC] Add drop processor (#5767)
opensearch-project/documentation-website: _ingest-pipelines/processors/drop.md

---
layout: default
title: Drop
parent: Ingest processors
nav_order: 70
---

# Drop processor

> The `drop` processor is used to discard documents without indexing them. This is useful for preventing documents from being indexed based on certain conditions. For example, you might use a `drop` processor to prevent documents that are missing important fields or contain sensitive information from being indexed. 

Reviewer: ```suggestion
The `drop` processor is used to discard documents without indexing them. This can be useful for preventing documents from being indexed based on certain conditions. For example, you might use a `drop` processor to prevent documents that are missing important fields or contain sensitive information from being indexed. 
```

The `drop` processor does not raise any errors when it discards documents, making it useful for preventing indexing problems without cluttering your OpenSearch logs with error messages.

## Syntax example

The following is the syntax for the `drop` processor:

```json
{
  "drop": {
    "if": "ctx.foo == 'bar'"
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `drop` processor.

Parameter | Required | Description |
|-----------|-----------|-----------|
`description`  | Optional  | A brief description of the processor.  |

> `if` | Optional | A condition for running this processor. |

Reviewer: ```suggestion
`if` | Optional | A condition for running the processor. |
```

`ignore_failure` | Optional | If set to `true`, failures are ignored. Default is `false`. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/) for more information. |
`on_failure` | Optional | A list of processors to run if the processor fails. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/) for more information. |

> `tag` | Optional | An identifier tag for the processor. Useful for debugging to distinguish between processors of the same type. |

Reviewer: ```suggestion
`tag` | Optional | An identifier tag for the processor. Useful for distinguishing between processors of the same type when debugging. |
```

## Using the processor

Follow these steps to use the processor in a pipeline.

> **Step 1: Create a pipeline.**

Reviewer: Global: I believe that these do not take periods.

The following query creates a pipeline, named `drop-pii`, that uses the `drop` processor to prevent a document containing personally identifiable information (PII) from being indexed:

```json
PUT /_ingest/pipeline/drop-pii
{
  "description": "Pipeline that prevents PII from being indexed",
  "processors": [
    {
      "drop": {
        "if" : "ctx.user_info.contains('password') || ctx.user_info.contains('credit card')"
      }
    }
  ]
}
```
{% include copy-curl.html %}

**Step 2 (Optional): Test the pipeline.**

> It is recommended that you test your pipeline before you ingest documents.

Reviewer: ```suggestion
It is recommended that you test your pipeline before ingesting documents.
```

{: .tip}

To test the pipeline, run the following query:

```json
POST _ingest/pipeline/drop-pii/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "user_info": "Sensitive information including credit card"
      }
    }
  ]
}
```
{% include copy-curl.html %}

#### Response

> The following example response confirms that the pipeline is working as expected, that is, the document has been dropped:

Reviewer: ```suggestion
The following example response confirms that the pipeline is working as expected (the document has been dropped):
```

```json
{
  "docs": [
    null
  ]
}
```
{% include copy-curl.html %}

**Step 3: Ingest a document.**

The following query ingests a document into an index named `testindex1`:

```json
PUT testindex1/_doc/1?pipeline=drop-pii
{
  "user_info": "Sensitive information including credit card"
}
```
{% include copy-curl.html %}

> The following response confirms that the document with ID `1` was not indexed: 

Reviewer: "with the ID `1`" or "with the ID of `1`"?

## Using the processor

Follow these steps to use the processor in a pipeline.

**Step 1: Create a pipeline.**

The following query creates a pipeline, named `drop-user`, that uses the `drop` processor to prevent a document containing personally identifiable information (PII) from being indexed:

```json
PUT /_ingest/pipeline/drop-pii
{
  "description": "Pipeline that prevents PII from being indexed",
  "processors": [
    {
      "drop": {
        "if" : "ctx.user_info.contains('password') || ctx.user_info.contains('credit_card')"
      }
    }
  ]
}
```
{% include copy-curl.html %}

**Step 2 (Optional): Test the pipeline.**

It is recommended that you test your pipeline before you ingest documents.
{: .tip}

To test the pipeline, run the following query:

```json
POST _ingest/pipeline/drop-pii/_simulate
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source": {
        "user_info": "Sensitive information including credit card"
      }
    }
  ]
}
```
{% include copy-curl.html %}

#### Response

The following example response confirms that the pipeline is working as expected:

```json
{

>   "docs": [

Reviewer 2: If the document was returned in the response, it means that the `drop` behavior didn't occur. In this case, I believe it's because of the mismatch between the condition checking for `credit_card` (with an underscore), while the text in the `user_info` field contains `credit card` (with a space), so the drop condition doesn't get satisfied.

I would suggest changing the pipeline configuration to check for `credit card` (with a space), which should result in the document getting removed from the request altogether. (The `_simulate` response should come back as an empty array, I believe. Also, you shouldn't be able to `GET` the document at the end, because it should have been dropped.)

Author: @msfroh Thank you for this clarification. Please confirm the revisions to steps 1 and 3 accurately reflect the info you provided. Step 4 (retrieve the document) has been deleted.
